src/librustc_mir/build/scope.rs

```diff
@@ -602,58 +612,42 @@ fn build_scope_drops<'tcx>(cfg: &mut CFG<'tcx>,
-                cfg.terminate(previous_block,
-                              scope.id,
-                              previous_span,
-                              TerminatorKind::Drop {
-                                  value: previous_value,
-                                  target: cached_block,
-                                  unwind: None
-                              });
-                return last_drop_block.unwrap();
-            } else {
-                return cached_block;
-            }
-        } else {
-            let block = cfg.start_new_cleanup_block();
-            drop_data.cached_block = Some(block);
-            if let Some((previous_block, previous_span, previous_value)) = previous {
-                cfg.terminate(previous_block,
-                              scope.id,
-                              previous_span,
-                              TerminatorKind::Drop {
-                                  value: previous_value,
-                                  target: block,
-                                  unwind: None
-                              });
-            } else {
-                last_drop_block = Some(block);
-            }
-            previous = Some((block, drop_data.span, drop_data.value.clone()));
-        }
-    }
-
-    // Prepare the end target for this chain.
-    let mut target = target.unwrap_or_else(||{
-        let b = cfg.start_new_cleanup_block();
-        cfg.terminate(b, scope.id, DUMMY_SP, TerminatorKind::Resume); // TODO
-        b
-    });
+                             unit_temp: &Lvalue<'tcx>,
+                             scopes: &mut [Scope<'tcx>])
+                             -> BasicBlock
+{
+    assert!(scopes.len() >= 1);
+
+    // Build up the drops in **reverse** order. The end result will
+    // look like:
+    //
+    //    [drops[n]] -...-> [drops[0]] -> [Free] -> [scopes[..n-1]]
+    //    |                                    |
+    //    +------------------------------------+
+    //     code for scopes[n]
+    //
+    // The code in this function reads from right to left. At each
+    // point, we check for cached blocks representing the
+    // remainder. If everything is cached, we'll just walk right to
+    // left reading the cached results but never created anything.
+
+    // To start, translate scopes[1..].
+    let (scope, earlier_scopes) = scopes.split_last_mut().unwrap();
+    let mut target = if let Some(cached_block) = scope.cached_block {
+        cached_block
+    } else if earlier_scopes.is_empty() {
+        // Diverging from the root scope creates a RESUME terminator.
+        // FIXME what span to use here?
+        let resumeblk = cfg.start_new_cleanup_block();
+        cfg.terminate(resumeblk, scope.id, DUMMY_SP, TerminatorKind::Resume);
+        resumeblk
+    } else {
+        // Diverging from any other scope chains up to the previous scope.
+        build_diverge_scope(tcx, cfg, unit_temp, earlier_scopes)
```

Comment:
This will also needlessly recurse on empty scopes (which are quite common), which were previously ignored by the `filter(|s| !s.drops.is_empty() || s.free.is_some())`.